src/dbc_packet.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2021 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::{Dbc, Error, PublicKey, Result};
+use blsttc::SecretKeySet;
+use rand::Rng;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::convert::Into;
+
+/// Represents a public key and optional
+/// secret key, plus a random derivation index.
+#[derive(Clone, Deserialize, Serialize)]
+pub struct DerivedKeySet {
+    public_key: PublicKey,
+    derivation_index: [u8; 32], // Todo: make this an Fr with new blsttc.
+
+    #[serde(
+        serialize_with = "serialize_option_sks",
+        deserialize_with = "deserialize_option_sks"
+    )]
+    secret_key_set: Option<SecretKeySet>,
+}
+
+impl DerivedKeySet {
+    /// public_key getter
+    pub fn public_key(&self) -> &PublicKey {
+        &self.public_key
+    }
+
+    /// derivation_index getter
+    pub fn derivation_index(&self) -> &[u8; 32] {
+        &self.derivation_index
+    }
+
+    /// secret_key_set getter
+    pub fn secret_key_set(&self) -> &Option<SecretKeySet> {
+        &self.secret_key_set
+    }
+
+    /// derives a public key
+    pub fn derived_public_key(&self) -> PublicKey {
+        self.public_key.derive_child(&self.derivation_index)
+    }
+
+    // generates a random derivation index
+    fn random_derivation_index() -> [u8; 32] {
+        rand::thread_rng().gen()
+    }
+}
+
+impl From<PublicKey> for DerivedKeySet {
+    /// construct a DerivedKeySet with random derivation index from a PublicKey
+    fn from(public_key: PublicKey) -> Self {
+        let derivation_index = Self::random_derivation_index();
+        Self {
+            public_key,
+            derivation_index,
+            secret_key_set: None,
+        }
+    }
+}
+
+impl From<SecretKeySet> for DerivedKeySet {
+    /// construct a DerivedKeySet with random derivation index from a SecretKeySet
+    fn from(secret_key_set: SecretKeySet) -> Self {
+        Self {
+            public_key: secret_key_set.public_keys().public_key(),
+            derivation_index: Self::random_derivation_index(),
+            secret_key_set: Some(secret_key_set),
+        }
+    }
+}
+
+type Memo = Option<[u8; 32]>;
```

Comment:
For a client only field, this size seems quite limiting. Can we relax this to Vec<u8>? perhaps with some (large) enforced cap?

Reply:
I went ahead and removed the field for now, per our conversation.  It can go into a future Payment struct.   I don't presently have a strong use-case example that requires a memo per DBC.